vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageRecorderView.kt

```diff
@@ -0,0 +1,227 @@
+
+    interface Callback {
+        fun onVoiceRecordingStarted()
+        fun onVoiceRecordingEnded()
+        fun onVoicePlaybackButtonClicked()
+        fun onVoiceRecordingCancelled()
+        fun onVoiceRecordingLocked()
+        fun onSendVoiceMessage()
+        fun onDeleteVoiceMessage()
+        fun onRecordingLimitReached()
+        fun onRecordingWaveformClicked()
+    }
+
+    // We need to define views as lateinit var to be able to check if initialized for the bug fix on api 21 and 22.
+    @Suppress("UNNECESSARY_LATEINIT")
+    private lateinit var voiceMessageViews: VoiceMessageViews
+    lateinit var callback: Callback
+
+    private var recordingTicker: CountUpTimer? = null
+    private var lastKnownState: RecordingUiState? = null
+    private var dragState: DraggingState = DraggingState.Ignored
+
+    init {
+        inflate(this.context, R.layout.view_voice_message_recorder, this)
+        val dimensionConverter = DimensionConverter(this.context.resources)
+        voiceMessageViews = VoiceMessageViews(
+                this.context.resources,
+                ViewVoiceMessageRecorderBinding.bind(this),
+                dimensionConverter
+        )
+        initListeners()
+    }
+
+    private fun initListeners() {
+        voiceMessageViews.start(object : VoiceMessageViews.Actions {
+            override fun onRequestRecording() = callback.onVoiceRecordingStarted()
+            override fun onMicButtonReleased() {
+                when (dragState) {
+                    DraggingState.Lock   -> {
+                        // do nothing,
+                        // onSendVoiceMessage, onDeleteVoiceMessage or onRecordingLimitReached will be triggered instead
+                    }
+                    DraggingState.Cancel -> callback.onVoiceRecordingCancelled()
+                    else                 -> callback.onVoiceRecordingEnded()
+                }
+            }
+
+            override fun onSendVoiceMessage() = callback.onSendVoiceMessage()
+            override fun onDeleteVoiceMessage() = callback.onDeleteVoiceMessage()
+            override fun onWaveformClicked() = callback.onRecordingWaveformClicked()
+            override fun onVoicePlaybackButtonClicked() = callback.onVoicePlaybackButtonClicked()
+            override fun onMicButtonDrag(nextDragStateCreator: (DraggingState) -> DraggingState) {
+                onDrag(dragState, newDragState = nextDragStateCreator(dragState))
+            }
+        })
+    }
+
+    override fun onVisibilityChanged(changedView: View, visibility: Int) {
+        super.onVisibilityChanged(changedView, visibility)
+        // onVisibilityChanged is called by constructor on api 21 and 22.
+        if (!this::voiceMessageViews.isInitialized) return
+        val parentChanged = changedView == this
+        voiceMessageViews.renderVisibilityChanged(parentChanged, visibility)
+    }
+
+    fun display(recordingState: RecordingUiState) {
```

Comment:
Nit: in the codebase those methods are named `render()`

Reply:
ah, will update in the other PR